File: androidApp/src/main/kotlin/com/rickroll/template/android/presentation/auth/AuthViewModel.kt
package com.rickroll.template.android.presentation.auth

import androidx.lifecycle.viewModelScope
import com.rickroll.template.android.presentation.mvi.MviViewModel
import com.rickroll.template.auth.domain.interactor.LoginInteractor
import com.rickroll.template.auth.domain.interactor.RegisterInteractor
import com.rickroll.template.auth.domain.model.AuthMode
import com.rickroll.template.core.result.RequestErrorCode
import com.rickroll.template.core.result.RequestResult
import kotlinx.coroutines.launch

class AuthViewModel(
    private val loginInteractor: LoginInteractor,
    private val registerInteractor: RegisterInteractor,
) : MviViewModel<AuthState, AuthEvent, AuthEffect>(AuthState()) {

    override fun handleEvent(event: AuthEvent) {
        when (event) {
            is AuthEvent.LoginChanged -> setState { copy(login = event.value) }
            is AuthEvent.PasswordChanged -> setState { copy(password = event.value) }
            is AuthEvent.ModeChanged -> setState { copy(mode = event.value, errorMessage = null) }
            AuthEvent.ErrorConsumed -> setState { copy(errorMessage = null) }
            AuthEvent.Submit -> submit()
        }
    }

    private fun submit() {
        withState { current ->
            viewModelScope.launch {
                setState { copy(isLoading = true, errorMessage = null) }

                val result = when (current.mode) {
                    AuthMode.LOGIN -> loginInteractor(current.login, current.password)
                    AuthMode.REGISTER -> registerInteractor(current.login, current.password)
                }

                when (result) {
                    is RequestResult.Success -> {
                        setState { copy(isLoading = false) }
                        sendEffect(AuthEffect.AuthSucceeded(current.mode))
                    }

                    is RequestResult.Error -> {
                        setState {
                            copy(
                                isLoading = false,
                                errorMessage = result.code.toUiMessage(),
                            )
                        }
                    }
                }
            }
        }
    }
}

private fun RequestErrorCode.toUiMessage(): String = when (this) {
    RequestErrorCode.EMPTY_LOGIN -> "Введите логин"
    RequestErrorCode.EMPTY_PASSWORD -> "Введите пароль"
    RequestErrorCode.LOGIN_ALREADY_EXISTS -> "Пользователь уже существует"
    RequestErrorCode.INVALID_CREDENTIALS -> "Неверный логин или пароль"
    RequestErrorCode.DATABASE,
    RequestErrorCode.UNKNOWN,
    -> "Ошибка авторизации"
}

File: iosApp/Sources/feature/auth/AuthViewModel.swift
import Foundation
import SharedKit
internal import Combine

struct AuthScreenState {
    var isAuthorized: Bool = false
    var isLoginMode: Bool = true
    var login: String = ""
    var password: String = ""
    var isLoading: Bool = false
    var errorMessage: String?
}

@MainActor
final class AuthViewModel: ObservableObject {
    @Published var state = AuthScreenState()

    private let authService = AuthService()

    func loadSession() {
        state.isAuthorized = authService.isAuthorized()
    }

    func switchMode(isLogin: Bool) {
        state.isLoginMode = isLogin
        state.errorMessage = nil
    }

    func submit() {
        state.isLoading = true
        state.errorMessage = nil

        let errorCode = state.isLoginMode
            ? authService.login(login: state.login, password: state.password)
            : authService.register(login: state.login, password: state.password)

        state.isLoading = false
        if errorCode == nil {
            state.password = ""
            state.isAuthorized = true
        } else {
            state.errorMessage = Self.mapError(errorCode)
        }
    }

    func logout() {
        if authService.logout() {
            state.isAuthorized = false
        }
    }

    private static func mapError(_ error: RequestErrorCode?) -> String {
        switch error {
        case .emptyLogin:
            return "Введите логин"
        case .emptyPassword:
            return "Введите пароль"
        case .loginAlreadyExists:
            return "Пользователь уже существует"
        case .invalidCredentials:
            return "Неверный логин или пароль"
        default:
            return "Ошибка авторизации"
        }
    }
}
